Sources/Apollo/GraphQLPaginatedQueryWatcher.swift

```diff
@@ -0,0 +1,157 @@
+#if !COCOAPODS
+import ApolloAPI
+#endif
+import Foundation
+
+public typealias Cursor = String
+
+/// Handles pagination in the queue by managing multiple query watchers.
+final class GraphQLPaginatedQueryWatcher<Query: GraphQLQuery, T>: Cancellable {
+
+  public struct Page {
+    let hasNextPage: Bool
+    let endCursor: Cursor?
+  }
+
+  public struct DataResponse {
+    let allResponses: [T]
+    let mostRecent: T
+    let source: GraphQLResult<Query.Data>.Source
+  }
+
+  /// Given a page, create a query of the type this watcher is responsible for
+  public typealias CreatePageQuery = (Page) -> Query?
+
+  private typealias ResultHandler = (Result<GraphQLResult<Query.Data>, Error>) -> Void
+
+  private let client: any ApolloClientProtocol
+
+  var initialWatcher: GraphQLQueryWatcher<Query>?
+  private var subsequentWatchers: [GraphQLQueryWatcher<Query>] = []
+
+  private let createPageQuery: CreatePageQuery
+  private let nextPageTransform: (DataResponse) -> T
+
+  private var modelMap: [Cursor?: T] = [:]
+  private var cursorOrder: [Cursor?] = []
+  private var resultHandler: ResultHandler?
+  private var callbackQueue: DispatchQueue
+  private var page: Page?
+
+  /// Designated initializer
+  ///
+  /// - Parameters:
+  ///   - client: The client protocol to pass in
+  ///   - inititalCachePolicy: The preferred cache policy for the initlal page. Defaults to `returnCacheDataAndFetch`.
+  ///   - callbackQueue: The queue for response callbacks.
+  ///   - query: The query to watch
+  ///   - createPageQuery: A function which creates a new `Query` given some pagination information.
+  ///   - transform: Transforms the `Query.Data` to the intended model and extracts the `Page` from the `Data`.
+  ///   - nextPageTransform: A function which combines extant data with the new data from the next page
+  ///   - onReceiveResults: The callback function which returns changes or an error.
+  public init(
+    client: ApolloClientProtocol,
+    inititalCachePolicy: CachePolicy = .returnCacheDataAndFetch,
+    callbackQueue: DispatchQueue = .main,
+    query: Query,
+    createPageQuery: @escaping CreatePageQuery,
+    transform: @escaping (Query.Data) -> (T?, Page?)?,
+    nextPageTransform: @escaping (DataResponse) -> T,
+    onReceiveResults: @escaping (Result<T, Error>) -> Void
+  ) {
+    self.callbackQueue = callbackQueue
+    self.client = client
+    self.createPageQuery = createPageQuery
+    self.nextPageTransform = nextPageTransform
+
+    let resultHandler: ResultHandler = { [weak self] result in
+      guard let self else { return }
+      switch result {
+      case .failure(let error):
+        guard !error.wasCancelled else { return }
+        // Forward all errors aside from network cancellation errors
+        onReceiveResults(.failure(error))
+      case .success(let graphQLResult):
+        guard let data = graphQLResult.data,
```

Comment:
One thing I'm not clear on here, is how this would handle data for an existing entity being updated. It seems like your code here does handle that, but I haven't wrapped my head around how that works completely yet and if there are edge cases where this could be problematic.

Maybe I just need to spend some more time looking at this part.